packages/apollo-server-defer/src/execute.ts

```diff
@@ -0,0 +1,1308 @@
+import { GraphQLError, locatedError } from 'graphql/error';
+import invariant from 'graphql/jsutils/invariant';
+import isInvalid from 'graphql/jsutils/isInvalid';
+import isNullish from 'graphql/jsutils/isNullish';
+import memoize3 from 'graphql/jsutils/memoize3';
+import promiseForObject from 'graphql/jsutils/promiseForObject';
+import promiseReduce from 'graphql/jsutils/promiseReduce';
+import { ObjMap } from 'graphql/jsutils/ObjMap';
+import { getDirectiveValues } from 'graphql/execution/values';
+import {
+  isObjectType,
+  isAbstractType,
+  isLeafType,
+  isListType,
+  isNonNullType,
+} from 'graphql/type/definition';
+import {
+  GraphQLObjectType,
+  GraphQLOutputType,
+  GraphQLLeafType,
+  GraphQLAbstractType,
+  GraphQLFieldResolver,
+  GraphQLResolveInfo,
+  ResponsePath,
+  GraphQLList,
+} from 'graphql/type/definition';
+import { GraphQLSchema } from 'graphql/type/schema';
+import {
+  DocumentNode,
+  OperationDefinitionNode,
+  FieldNode,
+  FragmentSpreadNode,
+  InlineFragmentNode,
+  FragmentDefinitionNode,
+} from 'graphql/language/ast';
+import {
+  ExecutionResult,
+  responsePathAsArray,
+  addPath,
+  assertValidExecutionArguments,
+  buildExecutionContext,
+  getOperationRootType,
+  collectFields,
+  buildResolveInfo,
+  resolveFieldValueOrError,
+  getFieldDef,
+} from 'graphql/execution/execute';
+import GraphQLDeferDirective from './GraphQLDeferDirective';
+import { Observable, Subscriber, merge } from 'rxjs';
+
+/**
+ * Rewrite flow types in typescript
+ */
+export type MaybePromise<T> = Promise<T> | T;
+
+function isPromise(
+  maybePromise: MaybePromise<any>,
+): maybePromise is Promise<any> {
+  return maybePromise && typeof maybePromise.then === 'function';
+}
+
+/**
+ * Determines if a field should be deferred. @skip and @include has higher
+ * precedence than @defer.
+ */
+function shouldDeferNode(
+  exeContext: ExecutionContext,
+  node: FragmentSpreadNode | FieldNode | InlineFragmentNode,
+): boolean {
+  const defer = getDirectiveValues(
+    GraphQLDeferDirective,
+    node,
+    exeContext.variableValues,
+  );
+  return defer && defer.if !== false; // default value for "if" is true
+}
+
+/**
+ * Define a new type for patches that are sent as a result of using defer.
+ * Its is basically the same as ExecutionResult, except that it has a "path"
+ * field that keeps track of the where the patch is to be merged with the
+ * original result.
+ */
+export interface ExecutionPatchResult {
+  data?: Record<string, any>;
+  errors?: ReadonlyArray<GraphQLError>;
+  path: ReadonlyArray<string | number>;
+}
+
+/**
+ * Define a return type from execute() that is a wraps over the initial
+ * result that is returned from a deferred query. Alongside the initial
+ * response, an observable that will stream patches of deferred fields is
+ * returned.
+ */
+export interface DeferredExecutionResult {
+  initialResult: ExecutionResult;
+  deferredPatchesObservable: Observable<ExecutionPatchResult>;
```

Comment:
I'm honestly still confused by the difference between `Observable` and async iterators, but since async iterators are already part of `graphql-js` for supporting subscriptions, and async iterators recently reached stage 4 and will be part of ES2018, should we consider using them instead? See http://2ality.com/2016/10/asynchronous-iteration.html.

Reply:
`Observable` is still at stage 1 and doesn't seem to be going anywhere: https://github.com/tc39/proposal-observable
Maybe we can ask @benjamn what his thoughts on this are.

Reply:
Also, I'm a little afraid of taking a dependency on RxJS because we want the bundle size on the server to stay as small as possible for edge execution. And V8 may get more optimized execution for async iterators (no idea if that makes sense or what it would look like).

Reply:
It turns out async iterators are already available natively in Node 10 and recent versions of Chrome! They require some getting used to, but for-await-of and async generators are pretty powerful. See http://2ality.com/2018/04/async-iter-nodejs.html for examples.

Reply:
Thanks for pointing this out Martijn! As per James' comments, I think this will require some more thinking about the pros and cons of both approaches. In the meantime, I will look at the subscriptions implementation to get a better understanding.